server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java

```diff
@@ -0,0 +1,147 @@
+        int valueCount = in.readInt();
+        final Streamer<T> streamer = dataType.streamer();
+        for (int i = 0; i < valueCount; i++) {
+            samples.add(streamer.readValueFrom(in));
+        }
+    }
+
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeLong(this.sampleCount);
+        out.writeInt(this.nullCount);
+        out.writeLong(this.totalBytes);
+        out.writeInt(this.samples.size());
+        final Streamer<T> streamer = dataType.streamer();
+        for (int i = 0; i < samples.size(); i++) {
+            streamer.writeValueTo(out, samples.get(i));
+        }
+    }
+
+    public void add(T value, RamAccounting ramAccounting) {
+        sampleCount++;
+        if (value == null) {
+            nullCount++;
+        } else {
+            totalBytes += dataType.valueBytes(value) + Long.BYTES;
+            ramAccounting.addBytes(totalBytes);
+            samples.add(value);
+        }
+    }
+
+    public ColumnStats<T> toColumnStats() {
+        samples.sort(dataType);
+        return ColumnStats.fromSortedValues(samples, dataType, nullCount, sampleCount);
+    }
+
+    /**
+     * Merges two builders together.  If the combined sample count of both builders is greater
+     * than {@code maxSampleCount} then a sampled subset of both builders are used.  This
+     * method may mutate this builder.
+     */
+    @SuppressWarnings("unchecked")
+    public ColumnStatsBuilder<T> merge(Random random, int maxSampleCount, ColumnStatsBuilder<?> other) {
+        if (Objects.equals(this.dataType, other.dataType) == false) {
+            throw new IllegalArgumentException(
+                "Cannot merge column stats from different datatypes [" + this.dataType + ", " + other.dataType + "]");
+        }
+        ColumnStatsBuilder<T> typedOther = (ColumnStatsBuilder<T>) other;
+        if (this.sampleCount == 0) {
+            return typedOther;
+        }
+        if (other.sampleCount == 0) {
+            return this;
+        }
+        if (this.sampleCount + other.sampleCount < maxSampleCount) {
+            for (T value : typedOther.samples) {
+                this.add(value, RamAccounting.NO_ACCOUNTING);
+            }
+            this.sampleCount += other.nullCount;
+            this.nullCount += other.nullCount;
+            return this;
+        }
+        ColumnStatsBuilder<T> mergedStats = new ColumnStatsBuilder<>(this.dataType);
+        Sampler<T> sampler = new Sampler<>(random, this.samples, typedOther.samples);
+        for (int i = 0; i < maxSampleCount; i++) {
+            mergedStats.add(sampler.next(), RamAccounting.NO_ACCOUNTING);
+        }
+        mergedStats.nullCount = this.nullCount + other.nullCount;
+        mergedStats.sampleCount = this.sampleCount + other.sampleCount;
+        return mergedStats;
+    }
+
+    // https://ballsandbins.wordpress.com/2014/04/13/distributedparallel-reservoir-sampling/
+    private static class Sampler<T> {
```

Comment:
Any reason to extract this into a class instead of keeping it inline as before? The state is available as locals in the one spot it's used, and it has just a single method.

Reply:
Yeah that can stay inline, I think this started off being more complicated and then got simpler again without me re-inlining it.